
<html>
<head>
<title>July, 2004: Exception Handling in Perl with Exception::Class</title>
</head>

<body BGCOLOR="#ffffff" LINK="#0000ff" VLINK="#330066" ALINK="#ff0000" TEXT="#000000">
<!--Copyright &#169; The Perl Journal-->

<h1>Exception Handling in Perl with Exception::Class</h1>
<p><i>The Perl Journal</i> July, 2004</p>
<h2></h2>


<h3>By Dave Rolsky</h3>


<I>Dave is the coauthor of Embedding Perl in HTML with Mason (O'Reilly and Associates) and can be reached at autarch@urth.org.</I>

<hr>





<p>Perl 5's built-in exception-handling facilities are minimal, to say the least. Basically, they consist of <i>eval</i> blocks (<i>eval { ... }</i>) and the <i>$@ </i>variable. If you want something a little more full featured, you'll have to turn to CPAN. Surprisingly enough, this is one area of CPAN where the number of choices is not exactly overwhelming.</p>

<p>The three exception-related modules on CPAN that I think are most useful and well-documented are <i>Error.pm</i>, <i>Exception.pm</i>, and my own module, <i>Exception::Class</i>. The first two aim at providing a more sophisticated <i>try</i>/<i>catch</i> syntax for Perl. My module, <i>Exception::Class</i>, is solely aimed at making it easier to declare hierarchies of exception classes. <i>Error.pm</i> and <i>Exception.pm</i> do provide base classes for exception objects, but they do not do much to help you create your own exception classes.</p>
<h3>Declaring Exception Classes</h3>

<p>To declare an exception class with <i>Exception::Class</i>, you simply <i>use Exception::Class</i>:</p>

<blockquote>
<pre>
use Exception::Class ( 'My::Exception::Class' );

</pre>
</blockquote>



<p>The above code snippet magically creates the new class <i>My::Exception::Class</i>. If you don't say otherwise, all exception classes created by <i>Exception::Class</i> are subclasses of <i>Exception::Class::Base</i>, which provides a basic constructor and accessors.</p>

<p>Of course, the above example isn't all that exciting, since you could just as easily have done it this way:</p>

<blockquote>
<pre>
package My::Exception::Class;
use base 'Exception::Class::Base';

</pre>
</blockquote>

<p>and gotten exactly the same result.</p>

<p>But what if you want to declare multiple classes at once, maybe for use within an application or suite of modules? You could type those <i>package</i> and <i>use base </i>bits over and over, or you could let <i>Exception::Class</i> do it for you. Here's an example adapted from the Mason code base:</p>

<blockquote>
<pre>
use Exception::Class
( 'HTML::Mason::Exception' =&gt;
  { description =&gt; 'generic base class for all Mason exceptions' },
  'HTML::Mason::Exception::Compiler' =&gt;
  { isa =&gt; 'HTML::Mason::Exception',
  description =&gt; 'error thrown from the compiler' },

  'HTML::Mason::Exception::Compilation' =&gt;
  { isa =&gt; 'HTML::Mason::Exception',
  description =&gt; "error thrown in eval of the code for a component" },

  'HTML::Mason::Exception::Compilation::IncompatibleCompiler' =&gt;
  { isa =&gt; 'HTML::Mason::Exception::Compilation',
  description =&gt; "a component was compiled by a compiler/lexer"
           . " with incompatible options.  recompilation is needed" },
);

</pre>
</blockquote>



<p>This is an extremely abbreviated version of the declarations found in the <i>HTML::Mason::Exceptions</i> module, but it demonstrates one of the core features of <i>Exception::Class</i>: creating a hierarchy of exception classes from a simple declaration. Note that after each class name, we pass a hash reference containing various options describing the particular class we want created.</p>

<p>The <i>description</i> parameter is a generic description of that whole class of exceptions. If your code catches an exception of a class you didn't create, adding the description to a log message may be handy. The description is not required, however.</p>

<p>The <i>isa </i>parameter is used to declare the parent of a given class. In the last example, we create a hierarchy several levels deep. The base class of our hierarchy is <i>HTML::Mason::Exception</i>. Below that we have <i>HTML::Mason::Compiler</i> and <i>HTML::Mason::Compilation</i>. And then finally, we have <i>HTML::Mason::Exception::Compilation::IncompatibleCompiler</i>, which subclasses <i>HTML::Mason::Exception::Compilation</i>.</p>
<h3>How <i>isa</i> is Handled</h3>

<p><i>Exception::Class</i> does its best to handle whatever you throw at it in the "<i>isa</i>" parameter. You are not required to list your classes in any specific order, which means that that this works fine:</p>

<blockquote>
<pre>
use Exception::Class
( 'My::Exception::Class::Subclass::Deep' =&gt;
  { isa =&gt; ''My::Exception::Class::Subclass' },

  'My::Exception::Class::Subclass' =&gt;
  { isa =&gt; ''My::Exception::Class' },

  'My::Exception::Class',
);

</pre>
</blockquote>

<p>Also, if you want to use your own custom class as the base class, that also works:</p>

<blockquote>
<pre>
package My::Exception::Class;

# no need to inherit from Exception::Class::Base if you don't want
# to, but you can.

sub new {
 ...
}

...


# in another file

use My::Exception::Class;
use Exception::Class

( 'My::Exception::Class::Subclass' =&gt;
  { isa =&gt; 'My::Exception::Class' } );

</pre>
</blockquote>

<p>You do need to make sure that your base class is loaded before you have <i>Exception::Class</i> create subclasses.</p>
<h3>Aliases</h3>

<p>In using exception classes, I've found that they can acquire some rather long and unwieldy names. The aforementioned <i>HTML::Mason::Exception::Compilation::IncompatibleCompiler</i> class name  is a particularly grotesque example. Typing this more than once can overwhelm even the most iron-fingered typist, and besides that, it's incredibly easy to make a typo in the name. These typos won't be caught at compile time, so you only find out about them when that particular bit of code is executed.</p>

<p>Because of this, <i>Exception::Class</i> makes it easy to create an alias for the class, which is a function that raises the appropriate exception.</p>

<p>Instead of typing this:</p>

<blockquote>
<pre>
HTML::Mason::Exception::Compilation::IncompatibleCompiler-&gt;throw
 ( error =&gt; ... );

</pre>
</blockquote>

<p>we can type the much less finger-straining version:</p>

<blockquote>
<pre>
wrong_compiler_error ...;

</pre>
</blockquote>



<p>Of course, you need to tell <i>Exception::Class</i> to make an alias:</p>

<blockquote>
<pre>
use Exception::Class
( ...,

  'HTML::Mason::Exception::Compilation::IncompatibleCompiler' =&gt;
  { isa		=&gt; 'HTML::Mason::Exception::Compilation',
  alias		=&gt; 'wrong_compiler_error',
  description 	=&gt;
  'a component was compiled by a compiler/lexer'
  . ' with incompatible options.  recompilation is needed' },
);

</pre>
</blockquote>

<p>When an <i>alias</i> parameter is specified, <i>Exception::Class</i> creates a handy little subroutine in the caller:</p>

<blockquote>
<pre>
sub wrong_compiler_error {
 HTML::Mason::Exception::Compilation::IncompatibleCompiler-&gt;throw(@_);
}

</pre>
</blockquote>

<p>Because of the way Perl parses subroutines, if we write a call to that subroutine without parentheses, Perl will check that subroutine name at compile time. Isn't that nice?</p>

<p>Of course, if you define your exceptions in one package and want to use them from many others, you'll need to reexport those generated subroutines. More on this later when we look at the use of <i>Exception::Class</i> within an application.</p>
<h3>Fields</h3>

<p>You'll often find that you want to add an extra field or two to an exception. For example, let's say that you are writing a parser and, if you encounter some bad syntax, you want to throw an exception. It would be very nice to include the name of the file and line number in the file where that error occurs. This could make debugging much easier because you could display the offending line when run in debugging mode.</p>

<p><i>Exception::Class</i> allows you to declare that a particular exception class has one or more arbitrary fields. These fields can be set when throwing an exception and will be accessible later via an accessor with the same name as the field.</p>

<p>So let's define our exception class with some fields:</p>

<blockquote>
<pre>
use Exception::Class
 ( 'My::Parser::Exception::Syntax' =&gt;
   { alias  =&gt; 'syntax_error',
   fields =&gt; [ 'filename', 'line_number' ] } );

</pre>
</blockquote>

<p>Now when we throw the exception, we can set those fields:</p>

<blockquote>
<pre>
My::Parser::Exception::Syntax-&gt;throw
 ( error		=&gt; 'Expected a florble after a glorp',
   filename	=&gt; $filename,
   line_number	=&gt; $line_number );

</pre>
</blockquote>

<p>And, of course, we can pass those fields to our handy <i>alias</i> subroutine:</p>

<blockquote>
<pre>
syntax_error
 error		=&gt; 'Expected a florble after a glorp',
 filename		=&gt; $filename,
 line_number	=&gt; $line_number;

</pre>
</blockquote>

<p>Later, we can get those values by calling <i>$@-&gt;filename()</i> and <i>$@-&gt;line_number()</i>.</p>
<h3>Throwing and Catching Exceptions</h3>

<p>As I mentioned earlier, if you want true <i>try</i>/<i>catch </i>syntax in Perl, you should take a look at either <i>Error.pm</i> or <i>Exception.pm</i>. In older versions of Perl, using these modules could cause memory leaks because they made it very easy to create a nested closure. Some testing I did recently seems to indicate that this is no longer a problem with Perl 5.8.4, though nested closures aren't specifically mentioned in any of the changelogs for recent versions.</p>

<p>If you are using an older Perl, you are probably better off avoiding these modules because creating a memory leak with them is so easy.</p>

<p>Of course, you can do exception handling with just <i>eval</i> and <i>$@</i>:</p>

<blockquote>
<pre>
eval {
 open my $fh, "&gt;$file"
   or My::Exception::System-&gt;throw
      ( error =&gt; "Cannot open $file for writing: $!" );
 ... # write something to the file
};

if ($@) {
 log_error( $@-&gt;error );
 exit;
}

</pre>
</blockquote>

<p>But there are a couple of pitfalls in this example.</p>

<p>First of all, we can't assume that, just because <i>$@</i> is true, that it contains an exception object. If the Perl interpreter itself threw an exception, then <i>$@</i> would just be a plain string.</p>

<blockquote>
<pre>
eval {
 open my $fh, "&gt;$file"
   or My::Exception::System-&gt;throw
      ( error =&gt; "Cannot open $file for writing: $!" );
 print $fh 10 / $x;
 close $fh;
};

if ($@) {
 log_error( $@-&gt;error );
 exit;
}

</pre>
</blockquote>

<p>If <i>$x</i> happens to be 0, then <i>$@</i> will contain something like "Illegal division by zero at foo.pl line 20." If you try to call the <i>error()</i> method on <i>$@</i> now, your program will die with a really helpful message like this:</p>

<blockquote>
<pre>
Can't locate object method "error" via package "Illegal division by zero at foo.pl line 20."

</pre>
</blockquote>



<p>Talk about confusing!</p>

<p>So you need to ensure that <i>$@</i> is an object before calling methods on it. One way to do this is to set <i>$SIG{__DIE__}</i> to something like this:</p>

<blockquote>
<pre>
sub make_exception_an_object {
 my $e = shift;
 ref $e &amp;&amp; $e-&gt;can('rethrow')
   ? $e-&gt;rethrow
   : My::Exception::Generic-&gt;throw( error =&gt; $e );
}

</pre>
</blockquote>

<p>Of course, this has several problems of its own. First, setting <i>$SIG{__DIE__}</i> is very, very global. Every single module you use in your code will be affected by this, which could cause some strange results. Even worse, there's no guarantee that some other code won't set it to a different value.</p>

<p>To be pedantic, I will point out that the test for <i>ref $e</i> does not actually indicate that <i>$e</i> is an exception object because it could very well be a plain, unblessed reference. So, if you can't cleanly force <i>$@</i> to always be an exception object, you're stuck testing for this when handling it:</p>

<blockquote>
<pre>
use Scalar::Util qw(blessed);

eval {
 open my $fh, "&gt;$file"
   or My::Exception::System-&gt;throw
      ( error =&gt; "Cannot open $file for writing: $!" );
 print $fh 10 / $x;
 close $fh;
};

if ($@) {
 if ( blessed $@ &amp;&amp; $@-&gt;can('error') ) {
   log_error( $@-&gt;error );
 } else {
   log_error( $@ );
 }
 exit;
}

</pre>
</blockquote>

<p>But wait, there are more problems. We need to be very careful about using <i>$@</i> after calling <i>eval</i> because it can mysteriously disappear. Consider, if you will, the following:</p>

<blockquote>
<pre>
if ($@) {
 log_error($@);
 warn $@ if DEBUG;
}

</pre>
</blockquote>

<p>It's possible that you might turn on debugging, see the value of <i>$@</i> logged, and then not see it printed via <i>warn</i>. In fact, <i>warn</i> might say something like "Warning: something's wrong at foo.pl line 23." Helpful, huh?</p>

<p>The reason this happens is that the <i>$@</i> variable is a global across all packages, and it gets reset every time Perl enters an <i>eval</i> block. So, if our <i>log_error()</i> subroutine, or for that matter any code that it calls, uses an <i>eval</i> block, then <i>$@</i> gets reset before the call to <i>warn</i> on the following line.</p>

<p>Because of that, I strongly recommend this idiom:</p>

<blockquote>
<pre>
eval { ... };

my $e = $@;
if ($e) { ... }

</pre>
</blockquote>

<p>If you copy <i>$@</i> into another variable right away, you can guarantee it won't be lost.</p>

<p>Of course, you can always encapsulate this into a nice little subroutine:</p>

<blockquote>
<pre>
eval { ... };

handle_error();
sub handle_error {
 return unless $@;
 my $e = $@;
 ...
}

</pre>
</blockquote>

<p>You'll notice that I've consistently tested <i>$@</i> for truth rather than defined-ness. This is because <i>$@</i> is set to the empty string "" when it is empty, rather than <i>undef</i>. Of course, if I were really diligent, I would test for <i>length $@</i>, in case someone does something like <i>die 0</i>.</p>
<h3>Constructor Parameters</h3>

<p>The method usually used to create a new exception is the <i>throw()</i> method, which constructs a new exception and then immediately calls <i>die</i> with that exception. You can also call <i>new()</i> to create a new exception if you want to do something else with it.</p>

<p>The constructor accepts several parameters. The most important parameter is the message you want to store in the exception. This can be passed in either the <i>error</i> or <i>message</i> parameter. As a special case, if one and only one parameter is passed to the constructor, then this is assumed to be the <i>message</i> parameter. This is done to make it possible to write code like this:</p>

<blockquote>
<pre>
system_error "Cannot write to $file: $!";

</pre>
</blockquote>



<p>If no message is given, then the <i>$!</i> variable is used, so you could even just write:</p>

<blockquote>
<pre>
open my $fh, "$file" or system_error;

</pre>
</blockquote>



<p>I'm not sure this is an idiom I'd use myself, but some may like it.</p>

<p>By default, when an exception object's <i>as_string()</i> method is called, it returns an error message without a stack trace. If you want this method's return value to include a stack trace, you can set the <i>show_trace</i> parameter to a true value when creating the object. If you always want objects of a certain class to include a trace, this can be controlled through a class method called <i>Trace()</i>, which will be covered later.</p>

<p>And as was shown earlier, any fields defined for the subclass being thrown can be set by passing them to the constructor:</p>

<blockquote>
<pre>
My::Parser::Exception::Syntax-&gt;throw
 ( error		=&gt; 'Expected a florble after a glorp',
   filename	=&gt; $filename,
   line_number	=&gt; $line_number );

</pre>
</blockquote>
<h3>Exception Object Methods</h3>

<p>Once you've caught an exception object, you'll probably want to do something with it. Since <i>Exception::Class::Base</i> overloads stringification, you can always just treat an exception object as a string, which is handy for logging. It also means that an uncaught exception that causes a program to die will generate some sort of useful error message.</p>

<p>If you want to get at individual pieces of information, you can use the following methods:</p>



<ul>
  <li><i><b>message()</i>, <i>error()</i></b>. Both of these methods return the error/message associated with the exception. Note that this is not the same as calling <i>as_string()</i>, which may include a stack trace.</li>
  <li><i><b>pid()</i></b>. Returns the PID at the time the exception was thrown.</li>
  <li><i><b>uid()</i></b>. Returns the real user ID at the time the exception was thrown.</li>
  <li><i><b>gid()</i></b>. Returns the real group ID at the time the exception was thrown.</li>
  <li><i><b>euid()</i></b>. Returns the effective user ID at the time the exception was thrown.</li>
  <li><i><b>egid()</i></b>. Returns the effective group ID at the time the exception was thrown.</li>
  <li><i><b>time()</i></b>. Returns the time in seconds since the epoch at the time the exception was thrown.</li>
  <li><i><b>package()</i></b>. Returns the package from which the exception was thrown.</li>
  <li><i><b>file()</i></b>. Returns the file within which the exception was thrown.</li>
  <li><i><b>line()</i></b>. Returns the line where the exception was thrown.</li>
  <li><i><b>trace()</i></b>. Returns the <i>Devel::StackTrace</i> object associated with the object. This class also overloads stringification, so you probably don't need to worry too much about the methods it offers unless you want to walk through the stack trace frame by frame.</li>
  <li><i><b>description()</i></b>. This method returns the description associated with the exception's class when the class was created. This can also be called as a class method.</li>
  <li><i><b>as_string()</i></b>. Returns the error message in string form, something like what you'd expect from <i>die</i>. If the class or object is set to show traces, then the full trace is also included, and the result looks a lot like calling <i>Carp::confess()</i>.</li>
  <li><i><b>full_message()</i></b>. This method is called by the <i>as_string()</i> method to get the message for the exception object. By default, this is the same as calling the <i>message()</i> method but may be overridden by a subclass. An example of why this is useful is shown later.</li>
  </ul>

<p>If you want to rethrow an exception object, you can call the <i>rethrow()</i> method. This is basically syntactic sugar for <i>die $exception</i>. This does not change any of the object's attribute values. You can also call the <i>show_trace()</i> method with a Boolean parameter to set whether a stack trace is included when the <i>as_string()</i> method is called.</p>
<h3>Class Methods</h3>

<p>Besides the <i>throw()</i> and <i>new()</i> constructors, <i>Exception::Class::Base</i> also offers several class methods that let you set certain behaviors on a per-class basis. These methods make use of the <i>Class::Data::Inheritable</i> module. That means that a subclass inherits the value set for a parent until the method is called for the subclass, at which point the subclass's value becomes independent.</p>

<p>The most important of these methods is <i>Trace()</i>, which allows you to control whether or not a stack trace is included in the value of the <i>as_string()</i> method for an object of a class. If <i>Trace()</i> is set to a true value, then the class and its children will default to including a trace.</p>

<p>The other two methods, <i>NoRefs()</i> and <i>RespectOverload(),</i> also take Booleans. They default to true and false, respectively. These methods control aspects of how <i>Devel::StackTrace</i> stores and serializes arguments when generating a string representation of a stack trace, and the defaults are probably acceptable for most uses. See the <i>Devel::StackTrace</i> and <i>Exception::Class</i> documentation for further details.</p>
<h3>Overriding <i>as_string()</i> and <i>full_message()</i></h3>

<p>The default implementation of the <i>as_string()</i> method does not take into account any fields that might exist for a given exception class. If you want to include these when an object is stringified or when <i>as_string()</i> is called, the correct way to do this is to override <i>full_message()</i> instead of overriding <i>as_string()</i>.</p>

<p>By default, the <i>full_message()</i> method is equivalent to the <i>message()</i> method. If you recall the <i>My::Parser::Exception::Syntax</i> class shown earlier, we could override <i>full_message() </i>to include the file and line number where the syntax error occurred:</p>

<blockquote>
<pre>
sub full_message {
 my $self = shift;
 return $self-&gt;message . ' at' . $self-&gt;filename
    . ' line ' . $self-&gt;line_number;
}

</pre>
</blockquote>

<p>This is much easier than overriding <i>as_string()</i>, which you are encouraged not to do.</p>
<h3>Using <i>Exception::Class</i> in an Application</h3>

<p>If you are using <i>Exception::Class</i> in an application or a large suite of modules, you will probably want to declare all your exceptions in one module. Additionally, you may want to provide some utility functions related to your exceptions.</p>

<p>For example, you might want a function that provides an easy way of determining if you are looking at a specific subclass in your exception hierarchy. Here is an example drawn from my <i>Alzabo</i> module:</p>

<blockquote>
<pre>
sub isa_alzabo_exception {
 my ($err, $name) = @_;
 return unless defined $err;

 if ($name) {
   my $class = "Alzabo::Exception::$name";
   return blessed $err &amp;&amp; $err-&gt;isa($class);
 } else {
   return blessed $err &amp;&amp; $err-&gt;isa('Alzabo::Exception');
 }
}

</pre>
</blockquote>

<p>So, inside the <i>Alzabo</i> code, if I want to know if an exception is of a certain subclass, I can do this:</p>

<blockquote>
<pre>
eval { .. };
my $e = $@;

if ( isa_alzabo_exception( $err, 'System' ) ) {
 ...
} elsif ( isa_alzabo_exception( $e, 'Params' ) ) {
 ...
} elsif ( isa_alzabo_exception($e) ) {
 ...
} else {
 # something else entirely
 ...
}

</pre>
</blockquote>

<p>This function is exported by my <i>Alzabo::Exceptions</i> module whenever it is loaded. You may also want to reexport the functions created by <i>Exception::Class</i> when the <i>alias</i> parameter is used. This is easily done using the <i>Exporter</i> module:</p>

<blockquote>
<pre>
package Alzabo::Exceptions;

use base 'Exporter';
use Exception::Class

 ( ...,

   'Alzabo::Exception::Logic' =&gt;
   { isa =&gt; 'Alzabo::Exception',
   alias =&gt; 'logic_exception',
   },

   'Alzabo::Exception::NoSuchRow' =&gt;
   { isa =&gt; 'Alzabo::Exception',
   alias =&gt; 'no_such_row_exception',
   },

   ...,
 );

@Alzabo::Exceptions::EXPORT_OK = ( 'logic_exception', 'no_such_row_exception' );

</pre>
</blockquote>



<p>As long as you import these reexported aliases at compile time, you can call them without using parentheses, as was shown previously.</p>
<h3>Integration with <i>Error.pm</i></h3>

<p>If you want to use exception classes created by <i>Exception::Class</i> with the <i>try</i>/<i>catch</i> syntax provided by <i>Error.pm</i>, you'll need to add the following hack to your code at some point:</p>

<blockquote>
<pre>
push @Exception::Class::Base::ISA, 'Error'
  unless Exception::Class::Base-&gt;isa('Error');

</pre>
</blockquote>

<p>It's a nasty little hack, but it works.</p>
<h3>Integration with <i>Exception.pm</i></h3>

<p>Because of the way <i>Exception.pm</i> implements exception throwing and handling, I am not sure if integrating it with <i>Exception::Class</i> is possible. My brief experiments resulted in failure. If someone figures out how to get them to play nice together, please let me know so I can include this in the <i>Exception::Class</i> documentation.</p>
<h3>Practical Applications</h3>

<p>David Wheeler created a handy module called <i>Exception::Class::DBI</i>, which integrates with DBI's error handling very nicely.</p>

<p>If you're interested in seeing how other people are using <i>Exception::Class</i>, there are a number of modules on CPAN to look at. My own <i>Alzabo</i> module uses <i>Exception::Class</i>, as does <i>HTML::Mason</i>. David Wheeler's Bricolage, a full-fledged CMS application, uses it as well. See the <i>Bric::Util::Fault</i> class to start with. If you want to test code that throws exceptions, take a look at Adrian Howard's <i>Test::Exception</i> module.</p>


<p><b></b></p>
<p><b></b></p>
<p><b>TPJ</b></p>




</body>
</html>